posthog/heatmaps/sql.py

```diff
@@ -0,0 +1,163 @@
+    $viewport_height Int16,
+    -- some elements move when the page scrolls, others do not
+    $pointer_target_fixed Bool,
+    $current_url VARCHAR,
+    _timestamp DateTime
+) ENGINE = {engine}
+"""
+
+HEATMAPS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine("heatmaps", replication_scheme=ReplicationScheme.SHARDED)
+
+HEATMAPS_TABLE_SQL = lambda: (
+    HEATMAPS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(timestamp)
+    -- almost always this is being queried by
+    --   * team_id,
+    --   * date range,
+    --   * URL (maybe matching wild cards),
+    --   * width
+    -- we'll almost never query this by session id
+    -- so from least to most cardinality that's
+    ORDER BY (team_id, $viewport_width, $current_url, toDate(timestamp))
+-- I am purposefully not setting index granularity
+-- the default is 8192, and we will be loading a lot of data
+-- per query, we tend to copy this 512 around the place but
+-- i don't think it applies here
+-- SETTINGS index_granularity=512
+"""
+).format(
+    table_name=HEATMAPS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=HEATMAPS_DATA_TABLE_ENGINE(),
+)
+
+KAFKA_HEATMAPS_TABLE_SQL = lambda: KAFKA_HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="kafka_heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=kafka_engine(topic=KAFKA_CLICKHOUSE_HEATMAP_EVENTS),
+)
+
+HEATMAPS_TABLE_MV_SQL = (
+    lambda: """
+CREATE MATERIALIZED VIEW IF NOT EXISTS heatmaps_mv ON CLUSTER '{cluster}'
+TO {database}.{target_table}
+AS SELECT
+    $session_id,
+    team_id,
+    timestamp,
+    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    x,
+    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    y,
+    -- stored so that in future we can support other resolutions
+    scale_factor,
+    $viewport_width,
+    $viewport_height,
+    -- some elements move when the page scrolls, others do not
+    $pointer_target_fixed,
+    $current_url
+FROM {database}.kafka_heatmaps
+""".format(
+        target_table="writable_heatmaps",
+        cluster=settings.CLICKHOUSE_CLUSTER,
+        database=settings.CLICKHOUSE_DATABASE,
+    )
+)
+
+# Distributed engine tables are only created if CLICKHOUSE_REPLICATED
+
+# This table is responsible for writing to sharded_heatmaps based on a sharding key.
+WRITABLE_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="writable_heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=Distributed(
+        data_table=HEATMAPS_DATA_TABLE(),
+        # I don't think there's a great natural sharding key here
+        # we'll be querying for team data by url and date,
+        # so I _think_ this offers a reasonable spread of write load
+        # without needing to query too many shards
+        sharding_key="cityHash64(concat(toString(team_id), '-', $current_url, '-', toString(toDate(timestamp))))",
```

Comment:
I would probably use something more random here. If a team has a breakout reddit/hacker news day and everyone is browsing that page writes are going to get really hot on just one shard. I'd try to use something like `session_id` instead of url maybe? That means that sessions will be colocated which is good, but _every_ session isn't going to go hammer one instance.

```suggestion
        sharding_key="cityHash64(concat(toString(team_id), '-', $session_id, '-', toString(toDate(timestamp))))",
```

Reply:
Yeah, I was torn on what to use here. Since we'll query across sessions so we always have to hit multiple shards to load data... but I think this is true however we chop this... all data for one query on one shard === hot shards for this product, so let's not do that 👍